### PR TITLE
SPLAT-1198: migrate vSphere installation to CAPV

### PR DESCRIPTION
**Pull request by rvanderp3**

### CODE REVIEW

File: pkg/infrastructure/vsphere/clusterapi/clusterapi.go [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+			}
+		}
+	}
+	return nil
+}
+
+func (p Provider) PreProvision(ctx context.Context, in clusterapi.PreProvisionInput) error {
+	installConfig := in.InstallConfig
+	clusterID := &installconfig.ClusterID{InfraID: in.InfraID}
+	vcenterContexts := &vcentercontexts.VCenterContexts{}
```
Comment (jcpowermac) [inline pkg/infrastructure/vsphere/clusterapi/clusterapi.go:310]: @patrickdillon wdyt?

**Comment by jcpowermac** [comment]
just doing this to test and see if it works, we should merge regardless

/test altinfra-e2e-vsphere-capi-ovn

**Comment by patrickdillon** [comment]
> just doing this to test and see if it works, we should merge regardless
> 
> /test altinfra-e2e-vsphere-capi-ovn

ok, joe! :smile_cat:

File: pkg/infrastructure/vsphere/clusterapi/clusterapi.go [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+	defer f.Close()
+
+	opts := soap.Upload{
+		ContentLength: size,
+	}
+
+	return lease.Upload(ctx, item, f, opts)
+}
+
+func importRhcosOva(ctx context.Context, session *session.Session, folder *object.Folder, cachedImage, clusterID, tagID, diskProvisioningType string, failureDomain vsphere.FailureDomain) error {
```
Comment (patrickdillon) [inline pkg/infrastructure/vsphere/clusterapi/clusterapi.go:96]: This code might be a bit easier to follow by breaking these functions into their own files

File: pkg/asset/machines/clusterapi.go [hunk 3]
```diff
@@ -249,12 +256,43 @@ func (c *ClusterAPI) Generate(dependencies asset.Parents) error {
...
+			}
+		}
 
+		pool.Platform.VSphere = &mpool
+		// todo: jcallen: this needs to be fixed too. template per region-zone pair
+		templateName := clusterID.InfraID + "-rhcos"
+
+		c.FileList, err = vspherecapi.GenerateMachines(ctx, clusterID.InfraID, ic, &pool, templateName, "master", vsphereContexts)
+		if err != nil {
+			return errors.Wrap(err, "unable to generate CAPI machines for vSphere")
```
Comment (patrickdillon) [inline pkg/asset/machines/clusterapi.go:289]: nit: use `fmt.Errorf`

File: pkg/asset/manifests/clusterapi/cluster.go [hunk 4]
```diff
@@ -105,6 +110,12 @@ func (c *Cluster) Generate(dependencies asset.Parents) error {
 		if err != nil {
 			return errors.Wrap(err, "failed to generate Azure manifests")
 		}
+	case vsphereplatform.Name:
+		var err error
+		out, err = vsphere.GenerateClusterAssets(installConfig, clusterID)
+		if err != nil {
+			return errors.Wrap(err, "failed to generate vSphere manifests")
```
Comment (patrickdillon) [inline pkg/asset/manifests/clusterapi/cluster.go:117]: use `fmt.Errorf`

**Review by patrickdillon — COMMENTED** [review]
This looks really good, but let's discuss refactoring the use of parents

**Comment by jcpowermac** [comment]
@patrickdillon github is being weird and doesn't let me respond on that first comment

> Hm, using parents like this is weird to me. I am surprised to see that it is working and will admit I don't really understand what's going on. I would need to revisit the asset graph to understand.

> Is the reason you're using parents here, is because vcenterContexts is too vsphere specific that you didn't want to add it to PreProvisionInput? I think it would be fine to add it there, and use the parents passed in from the cluster asset.

tbh I am a little concerned that PreProvisionInput becoming unwieldy ;-)

> You could avoid the issue entirely, by adding vcenterContexts as a metadata field of the installconfig, similar to what we do with aws metadata: https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/aws/metadata.go I think this would be the best approach and not too big of change...?

@rvanderp3 wdyt? It probably wouldn't be that bad of a change (famous last words).

File: pkg/infrastructure/vsphere/clusterapi/clusterapi.go (hunk 2 above)
Comment (jcpowermac) [reply pkg/infrastructure/vsphere/clusterapi/clusterapi.go:96]: ack, will work on that

**Comment by jcpowermac** [comment]
@patrickdillon can you please take a look at the last commit, that's the implementation of using metadata.

**Comment by patrickdillon** [comment]
> @patrickdillon can you please take a look at the last commit, that's the implementation of using metadata.

Looks good to me. WDYT? I noticed your comment:

```
// todo: jcallen: change context.TODO()
			// todo: jcallen: this seems very dangerous to be caching
			// todo: jcallen: the networks so soon. This might not belong here.
			/*
```

**Comment by jcpowermac** [comment]
> > @patrickdillon can you please take a look at the last commit, that's the implementation of using metadata.
> 
> Looks good to me. WDYT? I noticed your comment:
> 
> ```
> // todo: jcallen: change context.TODO()
> 			// todo: jcallen: this seems very dangerous to be caching
> 			// todo: jcallen: the networks so soon. This might not belong here.
> 			/*
> ```

I removed that section, Networks are called when its needed.

**Comment by jcpowermac** [comment]
Hmm so when I looked at the cluster in vcenter it was up and running. And I tested this earlier with no issues. 



/test altinfra-e2e-vsphere-capi-ovn

**Comment by patrickdillon** [comment]
Hm looks like machine api operator is complaining there are no compute nodes

**Comment by jcpowermac** [comment]
@patrickdillon I think this is ready now. Yesterday's run I think was a flake. We plan on creating a periodic as soon as this merges.

**Comment by jcpowermac** [comment]
The job for capi the openshift-test keeps getting oomkilled

https://github.com/openshift/release/pull/48835

**Comment by jcpowermac** [comment]
the capi job is hitting retesting since its random
https://issues.redhat.com/browse/OCPBUGS-16118
https://issues.redhat.com/browse/OCPBUGS-17855

/test altinfra-e2e-vsphere-capi-ovn

**Comment by patrickdillon** [comment]
/approve
"This is a good idea"

**Pull request merged**